api/agent/protocol/cloudevent.go

```diff
@@ -0,0 +1,207 @@
+package protocol
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	"go.opencensus.io/trace"
+
+	"github.com/fnproject/fn/api/models"
+)
+
+// CloudEvent is the official JSON representation of a CloudEvent: https://github.com/cloudevents/spec/blob/master/serialization.md
+type CloudEvent struct {
+	CloudEventsVersion string                 `json:"cloudEventsVersion"`
+	EventID            string                 `json:"eventID"`
+	Source             string                 `json:"source"`
+	EventType          string                 `json:"eventType"`
+	EventTypeVersion   string                 `json:"eventTypeVersion"`
+	EventTime          time.Time              `json:"eventTime"` // TODO: ensure rfc3339 format
+	SchemaURL          string                 `json:"schemaURL"`
+	ContentType        string                 `json:"contentType"`
+	Extensions         map[string]interface{} `json:"extensions"`
+	Data               interface{}            `json:"data"` // from docs: the payload is encoded into a media format which is specified by the contentType attribute (e.g. application/json)
+}
+
+type cloudEventIn struct {
+	CloudEvent
+
+	// Deadline string `json:"deadline"`
+	// Protocol CallRequestHTTP `json:"protocol"`
+}
+
+// cloudEventOut the expected response from the function container
+type cloudEventOut struct {
+	CloudEvent
+
+	// Protocol    *CallResponseHTTP `json:"protocol,omitempty"`
+}
+
+// CloudEventProtocol converts stdin/stdout streams from HTTP into JSON format.
+type CloudEventProtocol struct {
+	// These are the container input streams, not the input from the request or the output for the response
+	in  io.Writer
+	out io.Reader
+}
+
+func (p *CloudEventProtocol) IsStreamable() bool {
+	return true
+}
+
+func (h *CloudEventProtocol) writeJSONToContainer(ci CallInfo) error {
+	buf := bufPool.Get().(*bytes.Buffer)
+	buf.Reset()
+	defer bufPool.Put(buf)
+
+	_, err := io.Copy(buf, ci.Input())
+	if err != nil {
+		return err
+	}
+
+	var in cloudEventIn
+	// fmt.Printf("is cloudevent? %v\n", ci.IsCloudEvent())
+	if ci.IsCloudEvent() {
+		// then it's already in the right format, let's parse it, then modify
+		err = json.Unmarshal(buf.Bytes(), &in)
+		if err != nil {
+			return fmt.Errorf("Invalid CloudEvent input. %v", err)
+		}
+	} else {
+		in = cloudEventIn{
+			CloudEvent: CloudEvent{
+				ContentType:        ci.ContentType(),
+				EventID:            ci.CallID(),
+				EventType:          "http",
+				CloudEventsVersion: "0.1",
+				Source:             ci.RequestURL(),
+			},
+		}
+		if buf.Len() == 0 {
+			// nada
+			// todo: should we leave as null, pass in empty string, omitempty or some default for the content type, eg: {} for json?
+		} else if ci.ContentType() == "application/json" {
+			d := map[string]interface{}{}
+			err = json.NewDecoder(buf).Decode(&d)
+			if err != nil {
+				return fmt.Errorf("Invalid json body with contentType 'application/json'. %v", err)
+			}
+			in.Data = d
```

Comment:
so we're copying the input to a buffer, decoding it and re-encoding it just to get rid of quoted strings in a json?

https://play.golang.org/p/HSk9Enz2Zp5

a note would be nice. not sure why we need all the special json treatment.

Reply:
i see that it's making the data line a string instead of an object. the intent of the cloud event seems to be to let the final thing processing the cloud event parse the data section as specified by the content type, parsing the data field as a string json blob (i.e. `json.NewDecoder(strings.NewReader(cloudEvent.Data)).Decode(&event)`) seems to be the intent, not for us to do that? a slippery slope of us doing special treatment for other content types no?

Reply:
i get that it saves an extra decode, i still am not sure it's worth it, the intent or the right thing to do (but you've been staring at the spec more than me, surely)

Reply:
anyway, i guess this is going to get taken back out to the barn and shot soon enough anyway along with everything else in the 'if !IsCloudEvent` so just like my parents you can ignore me

Reply:
I have found our answer I believe https://github.com/cloudevents/spec/blob/master/http-transport-binding.md#22-data-attribute

Reply:
The serialization page just changed, but the new page is says the same as the old. The "data" field is not an escaped string, it's the full JSON as part of the whole object: https://github.com/cloudevents/spec/blob/master/json-format.md#32-examples

Reply:
So it's not left up to the final thing.

Reply:
got it, thanks for link